### ISSUE TITLE
Don't copy strings with escape characters

### ISSUE THREAD

**Issue by jkeiser** [simdjson/simdjson#1470]
Right now, we copy all strings to the string buffer because we don't know whether they contain escape characters--even though escape characters are few and far between (especially in keys). This is necessary for the DOM, which doesn't assume the input string stays around, but not necessary for On Demand. If we were able to expose a boolean array marking which strings (or tokens) have backslashes in them, we could avoid a ton of copying, returning string_views into the original JSON instead of the string buffer.

It would also eliminate the need for On Demand's feature vs. speed tradeoff where it (by default) doesn't support key lookups for keys with backslashes in them. We could check the boolean alongside the key at a very small cost, and branch when we need to to match the escaped key.

**Comment by lemire** [comment]
@jkeiser Can you elaborate ?

Let me first focus on DOM...

In the DOM case (at least), we need a reference to the string, somehow. If we do not copy it, then we must refer to the original location. If we can trust that we "own" the original input, then we can use the insitu approach that many parsers do... we just use the input as your string buffer, unescaping in place if needed.

There is a scenario where you also own the input, but where it is immutable. That's the scenario you are thinking about, I believe.

Finally, there is the scenario where you don't own the input. So the users can quickly reuse the input memory location, or even delete it, and you have no control over that. That's the scenario simdjson works with.

The case for our approach is clear. The case of insitu is clear. But I'd like to know about the scenario you have in mind in the DOM scenario.


Regarding On Demand... I think that the case is clear... because while On Demand is 'live', you are expecting to at least have access to the input throughout... it won't just disappear from under you normally (well, it might if we did batch processing?).

Tentatively marking 0.9.

**Comment by jkeiser** [comment]
Yeah, this is for On Demand only. Even with document stream, you have the text of the document around while the document is being processed.

**Comment by lemire** [comment]
> Even with document stream, you have the text of the document around while the document is being processed. 

Which is an argument for moving document_stream to On Demand.

**Comment by jkeiser** [comment]
This is a definite optimization thing. I'll move to 1.0, but I don't think we'd block 1.0 for it either.

**Comment by lemire** [comment]
Note that we do not copy keys by default in the On Demand API. We only copy string values.

**Comment by jkeiser** [comment]
Yeah, the issue is much smaller here. In fact, the user can choose to get the raw bytes of the JSON string, if they want, by asking for the raw string (which puts them at the `"`). Combined with strncmp and the fact that they know the maximum number of valid characters in the string, this is often enough to handle things like comparison.

It *might* be worth having an API that just gets them the characters inside the quotes, in a string_view. I'm not sure we should go further than that though.

**Comment by lemire** [comment]
So I tried something in https://github.com/simdjson/simdjson/pull/2212 but that's too complicated.

@CarlosEduR's PR at https://github.com/simdjson/simdjson/pull/2211 is not quite correct because , but I think that's the sort of simplicity we want.

Here is my current view. We should have a simple/cheap inlined function (which can use NEON/SSE2) equivalent to the following...

```cpp
std::none_of(s.begin(), s.end(), 
                       [](char c) { 
                        return c == '\\'; });
```


One thing that @CarlosEduR did not have is that you can call `raw_json_token()` to get a std::string_view that spans the whole string including the quotes. So all we need to do is call  `raw_json_token()` , then have a fast function check that there is no backslash, and if so we avoid the copy.

Unfortunately the `std::none_of` thing won't work because compilers cannot be trusted to generate good code. But it is the idea.

**Comment by lemire** [comment]
I have high hopes for https://github.com/simdjson/simdjson/pull/2596
